### PR TITLE
[Alternative] Fix Chat Bug

### LINKED ISSUE #4761: Network Chat Missing
## Repro Steps

In all cases below notice there is a network button but no chat window

(1) Local host 
- launch game
- click 'host game'

(2) With Local Lobby 
- start a local lobby
- start a headed game
- connect to lobby and host a game

(3) Local Lobby + Bot
- start a local lobby and headless bot
- start a headed game
- connect to lobby and join a bot

### Screenshot of problem:

In this screenshot notice there is no chat:
Screenshot from 2019-03-13 14-19-01

---------

If anyone wants to jump in on this one, feel free. I'll try to look into this otherwise at some point before the 1.10 release.

### PR DESCRIPTION
**Pull request by RoiEXLab**
## Overview
This fixes a regression introduced in #4563 alternative to #4782
Fixes #4761

## Functional Changes
Bugfix.

## Manual Testing Performed
I verified the bug no longer occurs.

### CODE REVIEW

File: game-core/src/main/java/games/strategy/engine/framework/startup/ui/panels/main/MainPanel.java [hunk 1]
```diff
@@ -113,8 +114,8 @@ private void addChat() {
     remove(chatSplit);
     chatPanelHolder.removeAll();
     final ChatModel chat = chatPanelSupplier.get().orElse(null);
-    if (chat instanceof SetupPanel) {
-      chatPanelHolder.add((SetupPanel) chat, BorderLayout.CENTER);
+    if (chat instanceof ChatPanel) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/engine/framework/startup/ui/panels/main/MainPanel.java]: If someone were to ever replace `ChatPanel` with a new subclass, but one that still maintained the same contract and woudl work, this instanceof would break. That is particularly bad to consider a change in some other, far-away, location then broke code here. 

`instanceof`:
- generally considered to be brittle
- often indicates poor design and abstractions.
- Using type information for control flow is essentially attaching an undeclared boolean flag to the object type which the 'caller' code then owns. For example, instead of having a headless flag every 'caller' could do an `instanceof` check and decide for itself what the definition of 'headless' is. It's a significant violation of data encapsulation.


IMO the original bug here was in part because an inheritently`instanceof` check is used. From my view instanceof is best used where strictly necessary and is something you tend to want to reduce to zero.

Also because as mentioned, if the sub-class were changes this would break. AFAIK the latest best practice advice is to favor using [isAssignableFrom](https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#isAssignableFrom(java.lang.Class)) instead.

**Comment by DanVanAtta** [comment]
@RoiEXLab see my comment. That we could change the `ChatPanel` type in the future, perhaps to something that still maintains the same contract, eg: `AwesomeChatPanel`, would have this code break. Even if it does not break, we'd then have to start doing 'or' conditions to look for more types.

I don't think we can even use `isAssignableFrom` here because the headless type should be excluded.  The fact that we exclude sub-classes seems also to be a violation of one of the SOLID principles, [liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) where the child class does *not* maintain the same contract as its parent (and hence why we no-op if we detect a sub-class). 

I know it's still bad design to have an `isHeadless` method, but it's more explicit at least and fixes much of the data encapsulation problems. It's a pretty common fix AFAIK to instanceof checks, simply to put a 'flag' method on a common parent class that can be used. In this way the implementation then controls the definition of 'headless'. 

My bottom line is instanceof is brittle, this could easily break again, or become ugly if we did want to have more types. Because 'instanceof' check is brittle, it's a task to find them and replace them. I've taken passes in this codebase to do just that and often finds places with problems (like here, where the UI code is making the decision of how to render, what it even means to be a 'headless' chat model. It's amazing the UI code even knows about the 'headless' concept or even a chat model to begin with, let alone defining what should be an encapsulated property of the ChatModel).

TL;DR instanceof is hella bad; use with caution, just like the `writeObject` method or other similar low level language features.

**Comment by DanVanAtta** [comment]
> Even if it does not break, we'd then have to start doing 'or' conditions to look for more types.

This was probably not clear. It'd be pretty dirty to see:

```
(panel instanceof ChatPanel) || (panel instanceOf AwesomeChatPanel))
```

File: game-core/src/main/java/games/strategy/engine/framework/startup/ui/panels/main/MainPanel.java [hunk 2]
```diff
@@ -114,8 +114,8 @@ private void addChat() {
     remove(chatSplit);
     chatPanelHolder.removeAll();
     final ChatModel chat = chatPanelSupplier.get().orElse(null);
-    if (chat instanceof ChatPanel) {
-      chatPanelHolder.add((ChatPanel) chat, BorderLayout.CENTER);
+    if (chat instanceof Component) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/engine/framework/startup/ui/panels/main/MainPanel.java]: This still checks for an implementation detail. The API of `ChatModel` does not have anything in common with `Component`, it's implementation that `ChatModel` also happens to be a `Component` (which is also bad as we are making our UI component literally also a model).

**Comment by DanVanAtta** [comment]
@RoiEXLab just so we don't lose information to this pull request,  I created a wiki page of TripleA anti-patterns and technical debt:  https://github.com/triplea-game/triplea/wiki/TripleA-Anti-Patterns-&-Technical-Debt

If the page is reasonable we can merge it into dev docs or link to it from the dev docs. There probably should be something somewhere that tells TripleA maintainers, "hey, don't do this".

**Comment by DanVanAtta** [comment]
@RoiEXLab I don't think this is a direction of fix that can be saved or should be done, please stop working on this. Let's move forward in other directions, I needed this fix for testing after changing network controls. We have lots of other very important things to do.

**Comment by RoiEXLab** [comment]
@DanVanAtta See my comment in #4782

**Comment by DanVanAtta** [comment]
http://www.javapractices.com/topic/TopicAction.do?Id=31

> The instanceof operator can be used to call a method based explicitly on the class of some object, instead of implicitly using an overridden method and polymorphism. Thus, [...] programmers may mistakenly use instanceof where an overidden method is more appropriate. 

Extremely strong no to this, we should strive for zero `instanceof` and casting. Keeping the code simpler, type safe, is part and parcel to moving fast. Having a cast here, and no test to make sure we don't break this, is just as well will break again.  Said again, there is another to stop us from breaking this code. What if the type does change? What if we successfully make the model to no longer be its own view. This is not compile time safe, there might be no real good reason for someone to even to *know* to check here.

A method can be tracked down, you remove it, and the code does not compile anymore. You are guaranteed to fix everything. That let's you move fast, no tip-toe'ing because you're not sure if you grepp'ed the code enough times or if you manually testing everything that could go wrong. Fast development is when you can know the exact 'radius' of a change and be confident you understand it and that it will be tested to still work after modification.

**Comment by RoiEXLab** [comment]
@DanVanAtta I agree with your reasoning, however as long as code isn't "made worse" I don't have a problem with keeping the potential problematic code for a short amount of time until a better structure is found.
Interfaces in combination with generics and the resulting compile time checks are great, but I personally view generics in java as sort-of broken, because in my opinion type erasure is scam. So IMO there certainly are a few reasonable cases where casts and instanceof checks do actually make sense.

Thanks for discussing this topic in-depth with me. I believe we both have the same end-goal, just a different opinion on how the code should look before actually transforming it into its "final form" 👍

**Comment by DanVanAtta** [comment]
I panic slightly when thinking about how much we need to do, I think there is a lesson in focus here.

I'll note that some refactors *do* require the code to be made worse, sometimes far worse before it can be made better. It's the 'one-step back for two steps forward' problem. An example I can give that I saw, in a code base with 10 classes that were not cleanly designed dependency wise, a first step was to make all methods static. The code was *far* worse, but it was a necessary step to be able to cleanly see and group method dependencies together, the next iteration regrouped the methods together into 3 or 4 better designed classes. That went on to fix a slew of usage problems. From another point of view, sometimes it's a balance of 2 problems, 1 problem which cannot be fixed, and a 2nd problem that can be fixed. Trading one part of a problem that cannot be fixed for three parts of a problem that can be fixed, is sometimes the trade that must be made.
